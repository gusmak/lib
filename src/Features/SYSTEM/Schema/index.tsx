import { AppProvider } from 'Utils';
import Container from './Container';
import { SchemaContext } from './context';
import { SchemaServices } from './Services';
import type { CurrentWorkspace } from '../types';
import { Provider } from 'jotai';

export type SchemaSystemProps = SchemaServices & {
    currentWorkspace?: CurrentWorkspace;
};

export function SchemaSystem(props: SchemaSystemProps) {
    const { currentWorkspace, ...services } = props;

    return currentWorkspace?.id ? (
        <AppProvider>
            <Provider>
                <SchemaContext.Provider
                    value={{
                        services,
                        currentWorkspace,
                    }}
                >
                    <Container />
                </SchemaContext.Provider>
            </Provider>
        </AppProvider>
    ) : null;
}

export * from './types';
export * from './Services';
export { default as FieldView } from './components/FieldView';
export default SchemaSystem;
