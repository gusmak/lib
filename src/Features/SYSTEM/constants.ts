export const Constants = {
    CREATE_PATH: 'Create',
    EDIT_PATH: 'Edit',
    ADD_EXISTED_USER: 'AddExistedUser',
    VIEW_PATH: 'View',
    OTHER_PATH: '/*',
    CREATE_ALL_PATH: 'CreateAll',
    WORKSPACE_SHARING_PATH: 'WorkspaceSharing',
    WORKSPACE_SHARING_CONFIG_PATH: 'Config',
    DIRECTORY_PERMISSION: 'Permission',
    SYSTEM_DIRECTORY_PERMISSION: 'SystemDirectory/Permission',
    PERMISSION_CODE: {
        FULL_CONTROL: 31,
        MODIFY: 15,
        WRITE: 4,
        READ_AND_EXECUTE: 3,
        READ: 2,
        LIST_FOLDER_CONTENTS: 1,
    },
};
