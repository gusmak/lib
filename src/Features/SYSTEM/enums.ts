
export enum WorkspaceType {
    AXN = 'AXN',
    ADMIN = 'ADMIN',
    ACM = 'ACM',
}
