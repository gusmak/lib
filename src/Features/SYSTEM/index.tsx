export { default as RoleSystem, type Role, type RoleServices } from './Role';
export { default as RoleTagSystem, type RoleTag, type RoleTagServices } from './RoleTag';
export type { PagingQueryInput, ObjectTypeCode } from 'Features/types';
export { type SharingProps } from './Sharing/Types';
export { default as GroupSystem } from './Group';
export { default as UserSystem } from './User';
export { default as SchemaSystem } from './Schema';
export { default as WorkflowSystem } from './Workflow';
export { default as DirectorySystem } from './Directory';
export { default as SharingSystem } from './Sharing';
