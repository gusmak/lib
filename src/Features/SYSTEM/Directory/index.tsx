import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Route, Routes } from 'react-router';
import { makeStyles } from '@mui/styles';
import { Box, Tab } from '@mui/material';
import { TabPanel, TabContext, TabList } from '@mui/lab';
import { AppProvider } from 'Utils';
import { Page } from 'AWING';
import { Constants } from '../constants';
import { DirectoryContext } from './context';
import Permission from './Permission';
import Container from './Container';
import { BORDER_LIGHTGRAY } from './constants';
import type { ObjectTypeCode } from 'Features/types';
import type { DirectoryServices } from './Services';
import type { CurrentWorkspace } from '../types';

export type DirectorySystemProps = DirectoryServices & {
    parentDirectoryId: number;
    currentWorkspace?: CurrentWorkspace;
    objectTypeCodes?: ObjectTypeCode[];
};

const useFilterTreeViewStyles = makeStyles({
    root: {
        '& .MuiTabs-flexContainer': {
            borderBottom: BORDER_LIGHTGRAY,
        },
    },
});

const DirectorySystem = (props: DirectorySystemProps) => {
    const { t } = useTranslation();
    const { parentDirectoryId, currentWorkspace, objectTypeCodes, ...services } = props;
    const classes = useFilterTreeViewStyles();

    const [tabValue, setTabValue] = useState<string>('1');

    useEffect(() => {
        const currentPath = window.location.href;
        setTabValue(currentPath.includes(Constants.SYSTEM_DIRECTORY_PERMISSION) ? '2' : '1');
    }, []);

    const handleChange = (_e: React.SyntheticEvent, newValue: string) => {
        setTabValue(newValue);
    };

    return (
        <AppProvider>
            <DirectoryContext.Provider
                value={{
                    parentDirectoryId,
                    currentWorkspace,
                    objectTypeCodes,
                    services,
                }}
            >
                <Page caption={t('DirectoryManagement.Title')}>
                    <Box sx={{ width: '100%', typography: 'body1' }}>
                        <TabContext value={tabValue}>
                            <TabList onChange={handleChange} indicatorColor={'primary'} className={`${classes.root}`}>
                                <Tab style={{ fontWeight: 'bold' }} label={t('DirectoryManagement.DomainDirectoryTab')} value="1" />
                            </TabList>
                            <TabPanel value="1">
                                <Container />
                            </TabPanel>
                        </TabContext>
                    </Box>
                </Page>
                {/* Routes */}
                <Routes>
                    <Route
                        key={Constants.DIRECTORY_PERMISSION + '/:id/*'}
                        path={Constants.DIRECTORY_PERMISSION + '/:id/*'}
                        element={<Permission />}
                    />
                </Routes>
            </DirectoryContext.Provider>
        </AppProvider>
    );
};

export * from './types';
export * from './Services';
export default DirectorySystem;
