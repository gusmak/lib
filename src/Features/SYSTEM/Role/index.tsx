import { useState } from 'react';
import { Routes, Route } from 'react-router';
import { AppProvider } from 'Utils';
import { Constants } from '../constants';
import Container from './Container';
import CreateOrEdit from './CreateOrEdit';
import { RoleContext } from './context';
import { RoleServices } from './Services';
import type { RoleTagOptions } from './types';

export type RoleSystemProps = RoleServices;

export function RoleSystem(props: RoleSystemProps) {
    const [roleTagOptions, setRoleTagOptions] = useState<RoleTagOptions[]>([]);

    const paths = [
        {
            param: Constants.EDIT_PATH + '/:id',
            element: <CreateOrEdit />,
        },
        {
            param: Constants.CREATE_PATH,
            element: <CreateOrEdit />,
        },
    ];

    return (
        <AppProvider>
            <RoleContext.Provider
                value={{
                    services: props,
                    roleTagOptions,
                    setRoleTagOptions,
                }}
            >
                <Container />
                <Routes>
                    {paths.map((p) => (
                        <Route key={p.param} path={p.param} element={p.element} />
                    ))}
                </Routes>
            </RoleContext.Provider>
        </AppProvider>
    );
}

export * from './types';
export * from './Services';
export default RoleSystem;
