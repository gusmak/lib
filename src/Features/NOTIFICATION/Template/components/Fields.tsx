import { Grid, Typography, Box } from '@mui/material';
import { ExpandMore, ChevronRight } from '@mui/icons-material';
import { TreeItem, SimpleTreeView } from '@mui/x-tree-view';
import { type ConvertObjectDefinition } from 'Features/SYSTEM/Schema';
import TemplateFieldView from './TemplateFieldView';
import { useState } from 'react';

type FieldsProps = {
    treeDatas: ConvertObjectDefinition[];
    label: string;
};

export default function Fields(props: FieldsProps) {
    const { treeDatas, label } = props;
    const [expanded, setExpanded] = useState<string[]>([]);

    const handleToggle = (_event: any, nodeIds: string[]) => {
        setExpanded(nodeIds);
    };

    return (
        <Grid container>
            <SimpleTreeView
                aria-label="controlled"
                slots={{ collapseIcon: ExpandMore, expandIcon: ChevronRight }}
                expandedItems={expanded}
                onExpandedItemsChange={handleToggle}
                //KV: TODO: mới tạm fixed chuyển sang thư viện mới TREEVIEW, cần fixed lại 2 thuộc tính dưới đây sau
                // selected={selected}
                // onNodeSelect={handleSelect}
                style={{ width: '100%', paddingBottom: '16px' }}
                disableSelection={true}
            >
                <TreeItem
                    itemId={'root'}
                    label={
                        <div
                            style={{
                                display: 'flex',
                                flexDirection: 'row',
                                justifyContent: 'space-between',
                                alignItems: 'center',
                            }}
                        >
                            <Box
                                sx={{
                                    display: 'flex',
                                    alignItems: 'center',
                                }}
                            >
                                <Typography>{label}</Typography>
                            </Box>
                        </div>
                    }
                >
                    {treeDatas.map((fieldInfo: ConvertObjectDefinition, index: number) => (
                        <TemplateFieldView key={index} fieldInfo={fieldInfo} />
                    ))}
                </TreeItem>
            </SimpleTreeView>
        </Grid>
    );
}
