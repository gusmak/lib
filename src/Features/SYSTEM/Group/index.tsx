import Container from './container';
import Create from './Create';
import Edit from './Edit';
import { useState } from 'react';
import { RoleTagOptions } from '../Role';
import { Constants } from '../constants';
import { Route, Routes } from 'react-router';
import { GroupContext } from './context';
import { GroupServices } from './Services';
import AppProvider from 'Utils/AppProvider';

export type GroupSystemProps = GroupServices;

export default function GroupSystem(props: GroupSystemProps) {
    const [roleTagOptions, setRoleTagOptions] = useState<RoleTagOptions[]>([]);

    const paths = [
        {
            param: Constants.CREATE_PATH,
            element: <Create />,
        },
        {
            param: Constants.EDIT_PATH + '/:id',
            element: <Edit />,
        },
    ];

    return (
        <AppProvider>
            <GroupContext.Provider
                value={{
                    services: props,
                    roleTagOptions,
                    setRoleTagOptions,
                }}
            >
                <Container />
                <Routes>
                    {paths.map((p) => (
                        <Route key={p.param} path={`${p.param}/*`} element={p.element} />
                    ))}
                </Routes>
            </GroupContext.Provider>
        </AppProvider>
    );
}
