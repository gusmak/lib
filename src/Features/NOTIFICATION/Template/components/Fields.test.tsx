import { render, screen, fireEvent } from '@testing-library/react';
import Fields from './Fields';
import { type ConvertObjectDefinition } from 'Features/SYSTEM/Schema';

jest.mock('@mui/material', () => ({
    Grid: ({ children }: { children: React.ReactNode }) => <div data-testid="grid">{children}</div>,
    Typography: ({ children }: { children: React.ReactNode }) => <span>{children}</span>,
    Box: ({ children }: { children: React.ReactNode }) => <div data-testid="box">{children}</div>,
}));

jest.mock('@mui/x-tree-view', () => ({
    SimpleTreeView: ({ children, onExpandedItemsChange }: { children: React.ReactNode; onExpandedItemsChange: Function }) => (
        <div data-testid="tree-view" onClick={() => onExpandedItemsChange(null, ['test-id'])}>
            {children}
        </div>
    ),
    TreeItem: ({ label, children }: { label: React.ReactNode; children?: React.ReactNode }) => (
        <div data-testid="tree-item">
            <div data-testid="tree-label">{label}</div>
            <div data-testid="tree-children">{children}</div>
        </div>
    ),
}));

jest.mock('./TemplateFieldView', () => ({
    __esModule: true,
    default: ({ fieldInfo }: { fieldInfo: any }) => <div data-testid="field-view">{fieldInfo.fieldName}</div>,
}));

describe('Fields Component', () => {
    const mockTreeData: ConvertObjectDefinition[] = [
        {
            fieldPath: 'test.path1',
            fieldName: 'Test Field 1',
            isDisable: false,
            childs: [],
            fieldType: '',
            id: 0,
            objectTypeCode: '',
            // schemaDetails: [],
        },
        {
            fieldPath: 'test.path2',
            fieldName: 'Test Field 2',
            isDisable: false,
            childs: [],
            fieldType: '',
            id: 0,
            objectTypeCode: '',
            // schemaDetails: [],
        },
    ];

    it('renders with label and tree data', () => {
        render(<Fields treeDatas={mockTreeData} label="Test Label" />);

        expect(screen.getByText('Test Label')).toBeInTheDocument();
        expect(screen.getAllByTestId('field-view')).toHaveLength(2);
    });

    it('handles tree expansion', () => {
        const { container } = render(<Fields treeDatas={mockTreeData} label="Test Label" />);

        const treeView = screen.getByTestId('tree-view');
        fireEvent.click(treeView);

        expect(container).toMatchSnapshot();
    });

    it('renders empty tree', () => {
        render(<Fields treeDatas={[]} label="Empty Tree" />);

        expect(screen.getByText('Empty Tree')).toBeInTheDocument();
        expect(screen.queryByTestId('field-view')).not.toBeInTheDocument();
    });

    it('renders complex tree structure', () => {
        const complexData: ConvertObjectDefinition[] = [
            {
                fieldPath: 'test.path1',
                fieldName: 'Parent',
                isDisable: false,
                childs: [
                    {
                        fieldPath: 'test.path1.child',
                        fieldName: 'Child',
                        isDisable: false,
                        childs: [],
                        fieldType: '',
                        id: 0,
                        objectTypeCode: '',
                        // schemaDetails: [],
                    },
                ],
                fieldType: '',
                id: 0,
                objectTypeCode: '',
                // schemaDetails: [],
            },
        ];

        render(<Fields treeDatas={complexData} label="Complex Tree" />);

        expect(screen.getByText('Complex Tree')).toBeInTheDocument();
        expect(screen.getAllByTestId('field-view')).toHaveLength(1);
    });
});
