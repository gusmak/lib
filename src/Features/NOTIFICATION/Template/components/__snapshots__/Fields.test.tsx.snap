// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fields Component handles tree expansion 1`] = `
<div>
  <div
    data-testid="grid"
  >
    <div
      data-testid="tree-view"
    >
      <div
        data-testid="tree-item"
      >
        <div
          data-testid="tree-label"
        >
          <div
            style="display: flex; flex-direction: row; justify-content: space-between; align-items: center;"
          >
            <div
              data-testid="box"
            >
              <span>
                Test Label
              </span>
            </div>
          </div>
        </div>
        <div
          data-testid="tree-children"
        >
          <div
            data-testid="field-view"
          >
            Test Field 1
          </div>
          <div
            data-testid="field-view"
          >
            Test Field 2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
